File: frontend/miniapp/src/utils/ledger.ts
export type LedgerStatus = 'recording' | 'ended'

export type LedgerMember = {
  id: string
  nickname: string
  avatarUrl?: string
  createdAt: string
}

export type LedgerRecord = {
  id: string
  memberId: string
  type: 'income' | 'expense'
  amount: number
  note?: string
  createdAt: string
}

export type LedgerBook = {
  id: string
  name: string
  status: LedgerStatus
  createdAt: string
  updatedAt: string
  endedAt?: string
  members: LedgerMember[]
  records: LedgerRecord[]
}

const STORAGE_KEY = 'scorehub.ledger.books.v1'

function nowIso(): string {
  return new Date().toISOString()
}

function safeArray<T>(input: any): T[] {
  return Array.isArray(input) ? input : []
}

function normalizeLedger(raw: any): LedgerBook | null {
  if (!raw || !raw.id) return null
  const createdAt = String(raw.createdAt || nowIso())
  const updatedAt = String(raw.updatedAt || createdAt)
  return {
    id: String(raw.id),
    name: String(raw.name || ''),
    status: raw.status === 'ended' ? 'ended' : 'recording',
    createdAt,
    updatedAt,
    endedAt: raw.endedAt ? String(raw.endedAt) : undefined,
    members: safeArray<LedgerMember>(raw.members),
    records: safeArray<LedgerRecord>(raw.records),
  }
}

function readAll(): LedgerBook[] {
  const raw = uni.getStorageSync(STORAGE_KEY)
  if (!raw) return []
  try {
    const list = Array.isArray(raw) ? raw : typeof raw === 'string' ? JSON.parse(raw) : []
    if (!Array.isArray(list)) return []
    return list.map(normalizeLedger).filter(Boolean) as LedgerBook[]
  } catch (e) {
    return []
  }
}

function writeAll(list: LedgerBook[]) {
  uni.setStorageSync(STORAGE_KEY, list)
}

function newId(prefix: string): string {
  return `${prefix}-${Date.now()}-${Math.random().toString(16).slice(2, 8)}`
}

function defaultSelfMember(): LedgerMember {
  const raw = (uni.getStorageSync('user') as any) || {}
  const nickname = String(raw?.nickname || '').trim() || '我'
  const avatarUrl = String(raw?.avatarUrl || '').trim()
  return {
    id: newId('member'),
    nickname,
    avatarUrl,
    createdAt: nowIso(),
  }
}

function sortLedgers(items: LedgerBook[]): LedgerBook[] {
  return items.slice().sort((a, b) => String(b.updatedAt).localeCompare(String(a.updatedAt)))
}

export function listLedgers(): LedgerBook[] {
  return sortLedgers(readAll())
}

export function getLedger(id: string): LedgerBook | null {
  const items = readAll()
  return items.find((it) => it.id === id) || null
}

export function upsertLedger(ledger: LedgerBook): LedgerBook {
  const items = readAll()
  const idx = items.findIndex((it) => it.id === ledger.id)
  if (idx >= 0) items[idx] = ledger
  else items.unshift(ledger)
  writeAll(items)
  return ledger
}

export function createLedger(name: string): LedgerBook {
  const now = nowIso()
  const selfMember = defaultSelfMember()
  const ledger: LedgerBook = {
    id: newId('ledger'),
    name: String(name || '').trim() || '记账簿',
    status: 'recording',
    createdAt: now,
    updatedAt: now,
    members: [selfMember],
    records: [],
  }
  const items = readAll()
  items.unshift(ledger)
  writeAll(items)
  return ledger
}

export function endLedger(id: string): LedgerBook | null {
  const ledger = getLedger(id)
  if (!ledger) return null
  if (ledger.status === 'ended') return ledger
  const now = nowIso()
  ledger.status = 'ended'
  ledger.endedAt = now
  ledger.updatedAt = now
  return upsertLedger(ledger)
}

export function addLedgerMember(id: string, payload: { nickname: string; avatarUrl?: string }): LedgerBook | null {
  const ledger = getLedger(id)
  if (!ledger || ledger.status === 'ended') return ledger
  const nickname = String(payload.nickname || '').trim()
  if (!nickname) return ledger
  const now = nowIso()
  const member: LedgerMember = {
    id: newId('member'),
    nickname,
    avatarUrl: payload.avatarUrl ? String(payload.avatarUrl).trim() : '',
    createdAt: now,
  }
  ledger.members = [...ledger.members, member]
  ledger.updatedAt = now
  return upsertLedger(ledger)
}

export function addLedgerRecord(
  id: string,
  payload: { memberId: string; type: 'income' | 'expense'; amount: number; note?: string },
): LedgerBook | null {
  const ledger = getLedger(id)
  if (!ledger || ledger.status === 'ended') return ledger
  const amount = Number(payload.amount)
  if (!payload.memberId || !Number.isFinite(amount) || amount <= 0) return ledger
  const now = nowIso()
  const record: LedgerRecord = {
    id: newId('record'),
    memberId: payload.memberId,
    type: payload.type === 'income' ? 'income' : 'expense',
    amount,
    note: payload.note ? String(payload.note).trim() : '',
    createdAt: now,
  }
  ledger.records = [record, ...ledger.records]
  ledger.updatedAt = now
  return upsertLedger(ledger)
}

File: backend/internal/http/handlers/ledger.go
package handlers

import (
	"context"
	"encoding/json"
	"net/http"
	"strings"
	"time"

	"github.com/cloudwego/hertz/pkg/app"

	"scorehub/internal/http/middleware"
	"scorehub/internal/store"
)

type LedgerHandlers struct {
	st *store.Store
}

func NewLedgerHandlers(st *store.Store) *LedgerHandlers {
	return &LedgerHandlers{st: st}
}

type createLedgerRequest struct {
	Name string `json:"name"`
}

type updateLedgerRequest struct {
	Name string `json:"name"`
}

func (h *LedgerHandlers) CreateLedger(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}

	var req createLedgerRequest
	if body, err := c.Body(); err == nil && len(body) > 0 {
		if err := json.Unmarshal(body, &req); err != nil {
			writeError(c, http.StatusBadRequest, "bad_request", "invalid json")
			return
		}
	} else if err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "read body failed")
		return
	}

	name := strings.TrimSpace(req.Name)
	if name == "" {
		name = time.Now().Format("2006-01-02 15:04") + " 记账"
	}

	user, err := h.st.GetUserByID(ctx, uid)
	if err != nil {
		writeError(c, http.StatusUnauthorized, "unauthorized", "user not found")
		return
	}

	ledger, member, err := h.st.CreateLedger(ctx, user, name)
	if err != nil {
		writeError(c, http.StatusInternalServerError, "internal", "db error", err)
		return
	}

	c.JSON(http.StatusOK, map[string]any{
		"ledger": toLedgerDTO(ledger),
		"member": toLedgerMemberDTO(member),
	})
}

func (h *LedgerHandlers) ListLedgers(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}

	items, err := h.st.ListLedgersForUser(ctx, uid)
	if err != nil {
		writeError(c, http.StatusInternalServerError, "internal", "db error", err)
		return
	}

	var out []any
	for _, it := range items {
		out = append(out, map[string]any{
			"id":          it.LedgerID,
			"name":        it.Name,
			"createdAt":   it.StartTime,
			"updatedAt":   it.UpdatedAt,
			"status":      it.Status,
			"endedAt":     it.EndedAt,
			"memberCount": it.MemberCount,
			"recordCount": it.RecordCount,
		})
	}

	c.JSON(http.StatusOK, map[string]any{"items": out})
}

func (h *LedgerHandlers) GetLedgerDetail(ctx context.Context, c *app.RequestContext) {
	id := strings.TrimSpace(c.Param("id"))
	if id == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "id required")
		return
	}

	ledger, members, records, err := h.st.GetLedgerDetail(ctx, id)
	if err != nil {
		if err == store.ErrNotFound {
			writeError(c, http.StatusNotFound, "not_found", "ledger not found")
			return
		}
		writeError(c, http.StatusInternalServerError, "internal", "db error", err)
		return
	}

	var memOut []any
	for _, m := range members {
		memOut = append(memOut, toLedgerMemberDTO(m))
	}
	var recOut []any
	for _, r := range records {
		recOut = append(recOut, toLedgerRecordDTO(r))
	}

	c.JSON(http.StatusOK, map[string]any{
		"ledger":  toLedgerDTO(ledger),
		"members": memOut,
		"records": recOut,
	})
}

func (h *LedgerHandlers) UpdateLedger(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}
	id := strings.TrimSpace(c.Param("id"))
	if id == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "id required")
		return
	}

	var req updateLedgerRequest
	body, err := c.Body()
	if err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "read body failed")
		return
	}
	if err := json.Unmarshal(body, &req); err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "invalid json")
		return
	}
	name := strings.TrimSpace(req.Name)
	if name == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "name required")
		return
	}

	ledger, err := h.st.UpdateLedgerName(ctx, id, uid, name)
	if err != nil {
		switch err {
		case store.ErrForbidden:
			writeError(c, http.StatusForbidden, "forbidden", "no permission")
			return
		case store.ErrNotFound:
			writeError(c, http.StatusNotFound, "not_found", "ledger not found")
			return
		default:
			writeError(c, http.StatusInternalServerError, "internal", "db error", err)
			return
		}
	}

	c.JSON(http.StatusOK, map[string]any{"ledger": toLedgerDTO(ledger)})
}

func (h *LedgerHandlers) AddLedgerMember(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}
	id := strings.TrimSpace(c.Param("id"))
	if id == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "id required")
		return
	}

	var req joinScorebookRequest
	body, err := c.Body()
	if err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "read body failed")
		return
	}
	if err := json.Unmarshal(body, &req); err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "invalid json")
		return
	}

	m, err := h.st.AddLedgerMember(ctx, id, uid, strings.TrimSpace(req.Nickname), strings.TrimSpace(req.AvatarURL))
	if err != nil {
		switch err {
		case store.ErrForbidden:
			writeError(c, http.StatusForbidden, "forbidden", "no permission")
			return
		case store.ErrNotFound:
			writeError(c, http.StatusNotFound, "not_found", "ledger not found")
			return
		case store.ErrScorebookEnded:
			writeError(c, http.StatusBadRequest, "ended", "ledger ended")
			return
		case store.ErrInvalidArgument:
			writeError(c, http.StatusBadRequest, "bad_request", "invalid nickname")
			return
		default:
			writeError(c, http.StatusInternalServerError, "internal", "db error", err)
			return
		}
	}

	c.JSON(http.StatusOK, map[string]any{"member": toLedgerMemberDTO(m)})
}

type addLedgerRecordRequest struct {
	MemberID string  `json:"memberId"`
	Type     string  `json:"type"`
	Amount   float64 `json:"amount"`
	Note     string  `json:"note"`
}

type updateLedgerMemberRequest struct {
	Nickname  string `json:"nickname"`
	AvatarURL string `json:"avatarUrl"`
}

func (h *LedgerHandlers) AddLedgerRecord(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}
	id := strings.TrimSpace(c.Param("id"))
	if id == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "id required")
		return
	}

	var req addLedgerRecordRequest
	body, err := c.Body()
	if err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "read body failed")
		return
	}
	if err := json.Unmarshal(body, &req); err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "invalid json")
		return
	}
	if strings.TrimSpace(req.MemberID) == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "memberId required")
		return
	}

	amount := req.Amount
	if !validTwoDecimals(amount) {
		writeError(c, http.StatusBadRequest, "bad_request", "amount must have at most 2 decimals")
		return
	}
	t := strings.ToLower(strings.TrimSpace(req.Type))
	if t != "expense" && t != "income" {
		writeError(c, http.StatusBadRequest, "bad_request", "invalid type")
		return
	}

	r, err := h.st.AddLedgerRecord(ctx, id, uid, strings.TrimSpace(req.MemberID), t, amount, strings.TrimSpace(req.Note))
	if err != nil {
		switch err {
		case store.ErrForbidden:
			writeError(c, http.StatusForbidden, "forbidden", "no permission")
			return
		case store.ErrNotFound:
			writeError(c, http.StatusNotFound, "not_found", "ledger not found")
			return
		case store.ErrScorebookEnded:
			writeError(c, http.StatusBadRequest, "ended", "ledger ended")
			return
		case store.ErrInvalidArgument:
			writeError(c, http.StatusBadRequest, "bad_request", "invalid record")
			return
		default:
			writeError(c, http.StatusInternalServerError, "internal", "db error", err)
			return
		}
	}

	c.JSON(http.StatusOK, map[string]any{"record": toLedgerRecordDTO(r)})
}

func (h *LedgerHandlers) UpdateLedgerMember(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}
	ledgerID := strings.TrimSpace(c.Param("id"))
	memberID := strings.TrimSpace(c.Param("memberId"))
	if ledgerID == "" || memberID == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "id required")
		return
	}

	var req updateLedgerMemberRequest
	body, err := c.Body()
	if err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "read body failed")
		return
	}
	if err := json.Unmarshal(body, &req); err != nil {
		writeError(c, http.StatusBadRequest, "bad_request", "invalid json")
		return
	}

	m, err := h.st.UpdateLedgerMember(ctx, ledgerID, uid, memberID, strings.TrimSpace(req.Nickname), strings.TrimSpace(req.AvatarURL))
	if err != nil {
		switch err {
		case store.ErrForbidden:
			writeError(c, http.StatusForbidden, "forbidden", "no permission")
			return
		case store.ErrNotFound:
			writeError(c, http.StatusNotFound, "not_found", "ledger member not found")
			return
		case store.ErrScorebookEnded:
			writeError(c, http.StatusBadRequest, "ended", "ledger ended")
			return
		case store.ErrInvalidArgument:
			writeError(c, http.StatusBadRequest, "bad_request", "invalid nickname")
			return
		default:
			writeError(c, http.StatusInternalServerError, "internal", "db error", err)
			return
		}
	}

	c.JSON(http.StatusOK, map[string]any{"member": toLedgerMemberDTO(m)})
}

func (h *LedgerHandlers) EndLedger(ctx context.Context, c *app.RequestContext) {
	uid, ok := middleware.UserID(c)
	if !ok {
		writeError(c, http.StatusUnauthorized, "unauthorized", "missing user")
		return
	}
	id := strings.TrimSpace(c.Param("id"))
	if id == "" {
		writeError(c, http.StatusBadRequest, "bad_request", "id required")
		return
	}

	ledger, err := h.st.EndLedger(ctx, id, uid)
	if err != nil {
		switch err {
		case store.ErrForbidden:
			writeError(c, http.StatusForbidden, "forbidden", "no permission")
			return
		case store.ErrNotFound:
			writeError(c, http.StatusNotFound, "not_found", "ledger not found")
			return
		default:
			writeError(c, http.StatusInternalServerError, "internal", "db error", err)
			return
		}
	}

	c.JSON(http.StatusOK, map[string]any{"ledger": toLedgerDTO(ledger)})
}

func toLedgerDTO(sb store.Scorebook) map[string]any {
	return map[string]any{
		"id":        sb.ID,
		"name":      sb.Name,
		"createdAt": sb.StartTime,
		"updatedAt": sb.UpdatedAt,
		"status":    sb.Status,
		"endedAt":   sb.EndedAt,
	}
}

func toLedgerMemberDTO(m store.LedgerMember) map[string]any {
	return map[string]any{
		"id":        m.ID,
		"role":      m.Role,
		"nickname":  m.Nickname,
		"avatarUrl": m.AvatarURL,
		"score":     m.Score,
		"createdAt": m.CreatedAt,
		"updatedAt": m.UpdatedAt,
	}
}

func toLedgerRecordDTO(r store.LedgerRecord) map[string]any {
	return map[string]any{
		"id":           r.ID,
		"memberId":     r.MemberID,
		"fromMemberId": r.FromMemberID,
		"toMemberId":   r.ToMemberID,
		"type":         r.Type,
		"amount":       r.Amount,
		"note":         r.Note,
		"createdAt":    r.CreatedAt,
	}
}
